### PR TITLE
Fix wayland builds

### PR DESCRIPTION
**Pull request by andydotxyz**
This picks all the files out of the wayland source that should compile with `-tags wayland`

### CODE REVIEW

**Comment by andydotxyz** [comment]
This fixes the errors in #278 - Though I now have new build errors (see the ticket)

**Comment by pwaller** [comment]
@andydotxyz thanks for doing the legwork.

One quick request: could you rebase your changes rather than introducing a merge?

Other than that - the only other thing I can think of it would be great to have is confirmation from someone else that it works for them. If you can provide me with a method of testing which would show that this works, then I'll do that to test it when I get a moment. Then we can merge. Patch looks reasonable on the face of it.

Please keep pinging this thread if you don't hear back in the near future (once a week is a good benchmark).

**Comment by danfragoso** [comment]
I was having the same problem when building with `-tags wayland`. I can confirm that @andydotxyz patch fixed the issue.

**Comment by andydotxyz** [comment]
Apologies I had to force push a couple of times as the conflicts kept kicking me.
This is now based on current master with no merges on the PR history.

**Comment by andydotxyz** [comment]
With regards to testing... using any Go-GLFW app `go run -tags wayland .` - this didn't compile or run before this patch, but once applied it will :).

File: scripts/generate-wayland-protocols.sh [hunk 1]
```diff
@@ -11,18 +11,29 @@ fi
...
-
-wayland-scanner code $TMP_CLONE_DIR/unstable/relative-pointer/relative-pointer-unstable-v1.xml "$GLGLFW_PATH"/wayland-relative-pointer-unstable-v1-client-protocol.c
-wayland-scanner client-header $TMP_CLONE_DIR/unstable/relative-pointer/relative-pointer-unstable-v1.xml "$GLGLFW_PATH"/wayland-relative-pointer-unstable-v1-client-protocol.h
-
-wayland-scanner code $TMP_CLONE_DIR/unstable/idle-inhibit/idle-inhibit-unstable-v1.xml "$GLGLFW_PATH"/wayland-idle-inhibit-unstable-v1-client-protocol.c
-wayland-scanner client-header $TMP_CLONE_DIR/unstable/idle-inhibit/idle-inhibit-unstable-v1.xml "$GLGLFW_PATH"/wayland-idle-inhibit-unstable-v1-client-protocol.h
+generate() {
+  HEADER=$1
+  VER=$2
+  if [[ "$VER" == "stable" ]]; then
```
Comment (pwaller) [inline scripts/generate-wayland-protocols.sh]: Minor issue here: On Ubuntu 20.04, /bin/sh is dash, so this fails because [[ is a bashism :(.

**Comment by pwaller** [comment]
Confirmed that it builds for me. Also confirmed that I get the same generated sources when I regenerate them. Minor issue introduced into generate-wayland-protocols.sh. I'm unable to test functionality but I trust that this is an improvement on that front.

I do see a warning:
```
# github.com/go-gl/glfw/v3.3/glfw
In file included from ./c_glfw_lin.go:9:
./glfw/src/wl_window.c:29: warning: "_GNU_SOURCE" redefined
   29 | #define _GNU_SOURCE
      | 
<command-line>: note: this is the location of the previous definition
```

It seems the issue is that we supply `-D_GNU_SOURCE` and additionally `wl_window.c` from upstream defines it.

Unfortunately if I get rid of the double define, the first `#include` of the internal headers in an earlier source file results in the wrong symbols being defined by the time wl_window. So either:
1. we should move wl_window to its own compilation unit, 
2. we should #undef and #redef _GNU_SOURCE on either side of `#include "wl_window.c".

A bit grim. But this is non-fatal and the situation is generally improved with this PR, so I think we should accept it at this point once generate-wayland-protocols.sh is fixed. Patches welcome to remove the warning also but not required.

**Comment by andydotxyz** [comment]
The origincal script did not work in dash either, due to the "==". But now it does.

**Comment by andydotxyz** [comment]
The _GNU_SOURCE issue seems to reate to https://github.com/go-gl/glfw/commit/bac7f0e2e1cc1b17dfc90d2c83c04f174e53acf8 
and I don't have a good solution sorry. I'd be happy for this to merge now as you say it can be improved later - at least it compile now!

**Comment by pwaller** [comment]
Yes, though that actually hasn't been merged (note the message saying the commit doesn't appear on any branch in this repo). That was me in #271 making a script to automatically fetch from the sources, and finding that someone had previously deleted it from the upstream sources (in v3.2). It still stands in v3.3 afaict.

And here it is upstream: https://github.com/glfw/glfw/blob/0ef149c8f2451fbc8e866834675a075cfc295b6c/src/wl_window.c#L29

**Comment by pwaller** [comment]
Thanks for your patience and following through with this :)

**Pull request merged**